node/libs/protobuf_build/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+
+        // Check that the compiled proto files belong to the declared proto package.
+        let package_root = self.proto_root.to_name();
+        for f in &descriptor.file {
+            if !package_root.contains(&f.package().into()) {
+                anyhow::bail!(
+                    "{:?} ({:?}) does not belong to package {:?}",
+                    f.package(),
+                    f.name(),
+                    package_root.to_string()
+                );
+            }
+        }
+
+        // Check that the compiled proto messages support canonical encoding.
+        canonical::check(&descriptor, &pool).context("canonical::check()")?;
+
+        // Prepare the output directory.
+        let output_dir = self
+            .input_root
+            .prepare_output_dir()
+            .context("prepare_output_dir()")?;
+        let output_path = output_dir.join("gen.rs");
+        let descriptor_path = output_dir.join("gen.binpb");
+        fs::write(&descriptor_path, &descriptor.encode_to_vec())?;
+        println!("PROTOBUF_DESCRIPTOR={descriptor_path:?}");
+
+        // Generate code out of compiled proto files.
+        let mut output = RustModule::default();
+        let mut config = prost_build::Config::new();
+        config.prost_path(self.this_crate().join("prost").to_string());
+        config.skip_protoc_run();
+        for d in &self.dependencies {
+            for f in &d.1.descriptor_proto.file {
+                let proto_rel = ProtoName::from(f.package())
+                    .relative_to(&d.1.proto_root)
+                    .unwrap();
+                let rust_abs = d.0.clone().join(proto_rel.to_rust_module());
+                config.extern_path(format!(".{}", f.package()), rust_abs.to_string());
+            }
+        }
+        let m = prost_build::Module::from_parts([""]);
+        for f in &descriptor.file {
+            let code = config
+                .generate(vec![(m.clone(), f.clone())])
+                .context("generation failed")?;
+            output
+                .sub(&ProtoName::from(f.package()).to_rust_module())
+                .append(&code[&m]);
+        }
+
+        // Generate the reflection code.
+        let output = output.sub(&self.proto_root.to_name().to_rust_module());
+        for proto_name in extract_message_names(&descriptor) {
+            output.append(&self.reflect_impl(proto_name));
+        }
+
+        // Generate the descriptor.
+        let rust_deps = self
+            .dependencies
+            .iter()
+            .map(|d| format!("&{}::DESCRIPTOR", d.0))
+            .collect::<Vec<_>>()
+            .join(",");
+        let this = self.this_crate().to_string();
+        output.append(&format!("\
+            pub static DESCRIPTOR : {this}::Lazy<{this}::Descriptor> = {this}::Lazy::new(|| {{\
+                {this}::Descriptor::new({:?}.into(), vec![{rust_deps}], &include_bytes!({descriptor_path:?}))\
+            }});\
+        ",package_root.to_string()));
```

Comment:
This looks fragile and error-prone. Ideally, we'd want to move as much code as possible to the crate itself, probably by using macro (since we need to use `include_bytes` macro), so that it's use in the generated code is as simple as possible, something like

```rust
{this}::declare_descriptor!({descriptor_path:?}, {package_root:?}, {rust_deps});
```

(here, `rust_deps` is a comma-separated list of `path`s [in terms of macros](https://doc.rust-lang.org/beta/reference/macros-by-example.html)).

Even more ideally, we'd want users to declare the descriptor themselves (cf. [`tonic::include_file_descriptor_set!`](https://docs.rs/tonic/latest/tonic/macro.include_file_descriptor_set.html)), but since we'd need to specify `rust_deps` in 2 places, this seems infeasible.

Reply:
made it a macro